### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by janga1997**
#808
Only one error I am not able to resolve.
`axelrod_first.py:442: error: "type" has no attribute "memory_depth"`

### CODE REVIEW

File: axelrod/player.py [hunk 1]
```diff
@@ -3,9 +3,10 @@
 import random
 import copy
 
-from axelrod.actions import Actions, flip_action
+from axelrod.actions import Actions, flip_action, Action
 from .game import DefaultGame
 
+from typing import Dict, Any
```
Comment (marcharper) [inline axelrod/player.py]: pedantic: alphabetize imports, move typing import above to the section with the standard library imports (this is in PEP8 IIRC)

File: axelrod/strategies/axelrod_first.py [hunk 2]
```diff
@@ -77,7 +79,7 @@ class RevisedDowning(Player):
         'manipulates_state': False
     }
 
-    def __init__(self, revised=True):
+    def __init__(self, revised: bool =True) -> None:
```
Comment (marcharper) [inline axelrod/strategies/axelrod_first.py]: space after `=`?

File: axelrod/strategies/gobymajority.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ class GoByMajority(Player):
     default this is 0)
     """
 
-    name = 'Go By Marjority'
+    name = 'Go By Majority'
```
Comment (marcharper) [inline axelrod/strategies/gobymajority.py]: jiminy cricket! are we not catching this with a test?

File: axelrod/strategies/memoryone.py [hunk 4]
```diff
@@ -1,7 +1,11 @@
-from axelrod.actions import Actions
+from axelrod.actions import Actions, Action
 from axelrod.player import Player
 from axelrod.random_ import random_choice
 
+from typing import List, Tuple, Union
+
+type_four_vector = Union[List[float], Tuple[float,float,float,float]]
```
Comment (marcharper) [inline axelrod/strategies/memoryone.py]: pedantic: spaces after commas

File: axelrod/strategies/titfortat.py [hunk 5]
```diff
@@ -487,7 +487,7 @@ class AdaptiveTitForTat(Player):
 
     Algorithm
 
-    if (opponent played C in the last cycle) then
+    if (opponent: Player  -> Actiplayed C in the last cycle) then
```
Comment (marcharper) [inline axelrod/strategies/titfortat.py]: something is weird with this line

File: axelrod/player.py (hunk 1 above)
Comment (janga1997) [inline axelrod/player.py]: I thought the typing imports should be kept separate from the library imports?

**Comment by janga1997** [comment]
@marcharper the only workaround I could find for the type error of `self.classifier['memory_depth']` is the declare the type of `classifier` as `Dict[str, Any]`, since the values in `classifier` are too varied to pin down to specific types.

**Comment by janga1997** [comment]
I didn't see the purpose of `__class__` in `self.__class__.memory_depth = rounds_to_cooperate` in `Tullock` in `axelrod_first`.
Changing it did not trip the unit test, but made mypy happy.

**Comment by janga1997** [comment]
Only one error in `sequence_player`.
`sequence_player.py:91: error: Too many arguments for "__init__" of "ThueMorse"`
I can't understand why it raises an error, since the it has the exact number of args the parent class
`SequencePlayer` requires.

File: axelrod/strategies/gobymajority.py (hunk 3 above)
Comment (drvinceknight) [inline axelrod/strategies/gobymajority.py]: Just taken a look: we kind of are catching it. The init method of the strategy renames the instance (which is tested):

```
 53         self.name = (                                                                                                                             
 54             'Go By Majority' + (self.memory > 0) * (": %i" % self.memory))                                                                        
 55         if self.soft:                                                                                                                             
 56             self.name = "Soft " + self.name                                                                                                       
 57         else:                                                                                                                                     
 58             self.name = "Hard " + self.name                                                                                                       
 59                                                  
```

 but yeah we can make that better:

```
 53         self.name = (                                                                                                                             
 54           self.name + (self.memory > 0) * (": %i" % self.memory))                                                                        
 55         if self.soft:                                                                                                                             
 56             self.name = "Soft " + self.name                                                                                                       
 57         else:                                                                                                                                     
 58             self.name = "Hard " + self.name                                                                                                       
 59                                                  
```

I think that would have tripped a test...

**Comment by janga1997** [comment]
Can someone have a look at this PR?
The rest of the files have too obscure variables for me to do in a sprint. I may have to tackle them file by file later.

File: axelrod/player.py (hunk 1 above)
Comment (marcharper) [inline axelrod/player.py]: It's not a big deal but usually standard library imports come before others. I guess we haven't discussed our official preference for typing imports.

File: axelrod/strategies/axelrod_first.py [hunk 6]
```diff
@@ -310,15 +312,16 @@ def __init__(self):
         super().__init__()
 
     @staticmethod
-    def score_history(my_history, opponent_history, score_map):
+    def score_history(my_history: List[Action], opponent_history: List[Action], score_map: Dict[Tuple[Action, Action], int]) -> int:
```
Comment (marcharper) [inline axelrod/strategies/axelrod_first.py]: Can you split into multiple lines:
```
def score_history(my_history: List[Action], opponent_history: List[Action],
                             score_map: Dict[Tuple[Action, Action], int]) -> int:
```

File: type_tests.sh [hunk 7]
```diff
@@ -38,3 +38,8 @@ mypy --ignore-missing-imports --follow-imports skip axelrod/strategies/oncebitte
 mypy --ignore-missing-imports --follow-imports skip axelrod/strategies/prober.py
 mypy --ignore-missing-imports --follow-imports skip axelrod/strategies/punisher.py
 mypy --ignore-missing-imports --follow-imports skip axelrod/strategies/rand.py
+mypy --ignore-missing-imports --follow-imports skip axelrod/strategies/memoryone.py
```
Comment (marcharper) [inline type_tests.sh]: Let's alphabetize these

**Comment by marcharper** [comment]
Can you check this out:

```
sh ./type_tests.sh

mypy: can't read file 'axelrod/strategies/worseandworse.py': No such file or directory
```

Should be "worse_and_worse.py". I'm not seeing the other error that you mentioned.

File: axelrod/strategies/axelrod_first.py (hunk 6 above)
Comment (janga1997) [inline axelrod/strategies/axelrod_first.py]: As for the imports, I guess I will have to make a change to all the files which have typehints.
Do you want me to tackle all of them, or just in any PRs from now?

**Comment by drvinceknight** [comment]
This looks good to me now.

**Comment by meatballs** [comment]
Default Value Syntax:

As @marcharper suggested, the guidance at http://mypy.readthedocs.io/en/latest/basics.html suggests that there should be spaces both before and after the '=' in arguments with default values. There are several instances here where one of those spaces is missing.

**Comment by janga1997** [comment]
@meatballs Made the changes.

**Comment by meatballs** [comment]
I'm happy with this one now but I'll leave it to @marcharper to decide if his requested changes are now ok.

**Pull request merged**